mungegithub/mungers/publish_scripts/util.sh

```diff
@@ -174,14 +628,111 @@ restore_vendor() {
         rm -rf ./vendor/github.com/google/gofuzz
+        # go-openapi is shared between apiserver and apimachinery
+        rm -rf ./vendor/github.com/go-openapi
     fi
-    git add --all
-    # check if there are new contents 
-    if git diff --cached --exit-code &>/dev/null; then
-        echo "vendor hasn't changed!"
-        return
+
+    git add Godeps/Godeps.json
+    git add vendor/ || true
+
+    # check if there are new contents
+    if git-index-clean; then
+        echo "Godeps.json hasn't changed!"
+    else
+        echo "Committing vendor/ and Godeps/Godeps.json."
+        git-with-user commit -q -m "sync: update godeps"
     fi
-    git -c user.name="Kubernetes Publisher" -c user.email="k8s-publish-robot@users.noreply.github.com" commit -m "sync: resync vendor folder"
+
+    # nothing should be left
+    ensure-clean-working-dir
+}
+
+# update the dependencies to the version checked out
+update-deps-in-godep-json() {
+    if [ ! -f Godeps/Godeps.json ]; then
+        return 0
+    fi
+
+    local is_library=${2}
+    local deps=${1}
+    local deps_array=()
+    IFS=',' read -a deps_array <<< "${1}"
+    local dep_count=${#deps_array[@]}
+    for (( i=0; i<${dep_count}; i++ )); do
+        local dep="${deps_array[i]%%:*}"
+        local dep_commit=$(cd ../${dep}; git rev-parse HEAD)
+        if [ -z "${dep_commit}" ]; then
+            echo "Couldn't find kube commit for current HEAD. Odd."
+            return 1
+        fi
+
+        # get old dependency hash in Godeps/Godeps.json
+        local old_dep_commit=$(jq -r '.Deps[] | select(.ImportPath | startswith("k8s.io/'${dep}'/") or . == "k8s.io/'${dep}'") | .Rev' Godeps/Godeps.json | tail -n 1)
+        if [ -n "${old_dep_commit}" ]; then
+            if [ "${old_dep_commit}" != "${dep_commit}" ]; then
+                # it existed before => replace with the new value
+                echo "Updating k8s.io/${dep} dependency to ${dep_commit}."
+                sed -i "s/${old_dep_commit}/${dep_commit}/g" Godeps/Godeps.json
+            fi
+        elif git grep -w -q "k8s.io/${dep}"; then
+            # revert changes and fall back to full vendoring
+            echo "Found new dependency k8s.io/${dep}. Switching to full vendoring."
+            git checkout -q HEAD Godeps/Godeps.json
+            update_full_godeps "${deps}" "${is_library}"
+            return $?
+        else
+            echo "Ignoring k8s.io/${dep} dependency because it seems not to be used."
+        fi
+    done
+
+    git add Godeps/Godeps.json
+
+    # check if there are new contents
+    if git-index-clean; then
+        echo "Godeps.json hasn't changed!"
+    else
+        echo "Committing Godeps/Godeps.json."
+        git-with-user commit -q -m "sync: update godeps"
+    fi
+
+    # nothing should be left
+    ensure-clean-working-dir
+}
+
+# checkout the dependencies to the versions corresponding to the kube commit of HEAD
+checkout-deps-to-kube-commit() {
```

Comment:
Can we just checkout the tip of the corresponding branch of the deps?

Reply:
The tip is not stable over time. Imagine a dependency gets another commit, but the current repo does not (because no commits modify files in its staging directory). Then the tip changes and the Godeps.json is updated when it shouldn't. In other words: we only want to update the dependencies revisions when the current repo changes as well.

Reply:
>  Imagine a dependency gets another commit, but the current repo does not (because no commits modify files in its staging directory). 

Even in this case, it's safe to update the current repo's Godeps.json, though it's not needed. The fact that the current repo and the dpendency repo co-exists in the kubernetes repo makes sure at any time, their tips work together.